calendar/api/parent/ext-calendar-provider.js

```diff
@@ -408,6 +449,21 @@ this.calendar_provider = class extends ExtensionAPI {
               };
             },
           }).api(),
+
+          onSend: new EventManager({
```

Comment:
Who's calling `onSend` and why? Please document, e.g.:

When we need to send out an invitation, the API calls the extension to do the actual sending, using the server's protocol (e.g. SMTP, Exchange, etc.).

The extension needs to implement `onSend` with the following parameters: ...

Reply:
(`onSend` isn't a function, it's an event (in extension terminology, not to be confused with DOM events or calendar events).)

If the extension wants to handle sending mail on behalf of the calendar, it can achieve this by registering a listener using `messenger.calendar.provider.onSend.addListener`. When Lightning wants to send an email, the experiment code will send an extension event, which calls the listener with three parameters, `calendar`, `mimeContent` and `recipients`. (`sendItems` is not provided with a true `calIEvent` object, so we can't be more helpful than that.) The extension is then expected to send the `mimeContent` as an email to the `recipients`. (Note that the `To` header is already populated in the `mimeContent`, so the message can be e.g. dropped into an SMTP pickup folder as-is.)

Reply:
> onSend isn't a function()

Sure. But somebody calls the event handler. I meant just to document this in the code with a comment, so that it's clear to a _casual_ reader what the overall control flow is.

Reply:
So IIUC, if there are no listeners to onSend, then the base provider class takes over and sends via email. If there is an onSend listener, then it gets called and the extension is expected to do the sending. I think we should provide a use case where the extension gets the details about sending and can then decide if it wants to do the sending itself, or defer to email. I'm imagining this to be a little like webRequest, where the returned object defines how things are handled, and in that case maybe onBeforeSend may make more sense.

Reply:
> But somebody calls the event handler.

Well, the experiment does. And somebody calls the experiment. Buyt that's already documented in Lightning. So I'm not sure what more you're asking for.

Reply:
@NeilRashbrook: Here's an example code comment that documents what onSend does and how it works. Basically what we're discussing here.

```
/**
When we need to send out an invitation, the API calls the calendar provider extension via this listener, allowing the extension to do the actual sending itself, using the server's protocol (e.g. SMTP, Exchange, etc.).

There are 3 ways to send the invitation:
* Let Lightning send the invitation, using SMTP and the associated identity in Thunderbird. This happens, if there are no `onSend` listeners attached, or if all listeners return nothing or return `.didSend = false`.
* The calendar provider extension sends the invitiation. For that purpose, the parameters to the listener include the event and a pre-constructed MIME message, which the extension can send using a custom server protocol, or it can construct the invitation on its own using the event data. It needs to return `.didSend = true` to tell Lightning that the sending is done.
*  The calendar provider knows that the server will send out or has already sent out the invitiation on its own, simply by adding the event to the calendar, so there's no need to send another invitation. In this case, the handler just needs to return `.didSend = true`.

The parameters passed to the listener:
...
 */
onSend: new EventManager({
```

Reply:
@NeilRashbrook : Please include the comment to document the API.

Reply:
@NeilRashbrook : Please include this code comment, to document the API.